Vic2ToHoI4/Source/HOI4World/Decisions/GenericDecisions.cpp

```diff
@@ -1,391 +1,376 @@
-			}
-			else if (decision.getName() == "blow_panama_canal")
-			{
-				category.replaceDecision(updateBlowPanama(std::move(decision), provinceToStateIdMap));
-			}
-			else if (decision.getName() == "rebuild_suez_canal")
-			{
-				category.replaceDecision(updateRebuildSuez(std::move(decision), provinceToStateIdMap));
-			}
-			else if (decision.getName() == "rebuild_panama_canal")
-			{
-				category.replaceDecision(updateRebuildPanama(std::move(decision), provinceToStateIdMap));
-			}
-			else if (decision.getName() == "women_in_the_workforce")
-			{
-				category.replaceDecision(updateWomenInTheWorkforce(std::move(decision), majorIdeologies));
-			}
-			else if (decision.getName() == "war_bonds")
-			{
-				category.replaceDecision(updateWarBonds(std::move(decision), majorIdeologies));
-			}
-			else if (decision.getName() == "improved_worker_conditions")
-			{
-				category.replaceDecision(updateImprovedWorkerConditions(std::move(decision)));
-			}
-			else if (decision.getName() == "promises_of_peace")
-			{
-				category.replaceDecision(updatePromisesOfPeace(std::move(decision)));
-			}
-			else if (decision.getName() == "war_propaganda")
-			{
-				category.replaceDecision(updateWarPropaganda(std::move(decision)));
-			}
-			else if (decision.getName() == "war_propaganda_against_warmonger")
-			{
-				category.replaceDecision(updateWarPropagandaAgainstWarmonger(std::move(decision)));
-			}
-			else if (decision.getName() == "object_to_attaches")
-			{
-				category.replaceDecision(updateObjectToAttaches(std::move(decision)));
-			}
-		}
+	return relevantStates;
+}
+
+decision&& updateBuildKiel(decision&& buildKielDecision, const std::map<int, int>& provinceToStateIdMap)
+{
+	const std::string canalStatesPlaceholder = "$CANAL_STATES";
+	auto relevantCanalStates = getRelevantStatesFromProvinces(kielCanalProvinces, {}, provinceToStateIdMap);
+
+	std::string canalFullControlStatesString;
+	for (const auto& state: relevantCanalStates)
+	{
+		canalFullControlStatesString += "has_full_control_of_state = " + std::to_string(state) + "\n\t\t\t";
 	}
+	canalFullControlStatesString = canalFullControlStatesString.substr(0, canalFullControlStatesString.length() - 4);
+
+	std::string removeEffect = buildKielDecision.getAvailable();
+	removeEffect.replace(removeEffect.find(canalStatesPlaceholder),
+		 canalStatesPlaceholder.size(),
+		 canalFullControlStatesString);
+	buildKielDecision.setAvailable(removeEffect);
 
-	decisions.erase(std::remove_if(decisions.begin(),
-							  decisions.end(),
-							  [](auto& decisionCategory) {
-								  return decisionCategory.getName() == "foreign_support";
-							  }),
-		 decisions.end());
+	return std::move(buildKielDecision);
 }
 
 
-std::set<int> getRelevantStatesFromProvinces(const std::set<int>& provinces,
-	 const std::set<int>& statesToExclude,
-	 const std::map<int, int>& provinceToStateIdMap);
+decision&& updateBuildSuez(decision&& buildSuezDecision, const std::map<int, int>& provinceToStateIdMap)
```

Comment:
This function is almost the same as above, maybe you could create a generic function for `updateBuildX`?